Use Conda for pre-commit checks

martyngigg opened the pull request:
**Note: When this is merged the the Jenkins job config will need to be updated. Ask someone from the DevOps group to take a look.**

**Description of work.**

The `pre-commit` job recently starting hitting this error:

```sh
An unexpected error has occurred: CalledProcessError: command: ('/home/abc/.cache/pre-commit/repoyrc1wg3z/py_env-python3/bin/python', '-mpip', 'install', '.')
return code: 1
expected return code: 0
stdout:
    Processing /home/abc/.cache/pre-commit/repoyrc1wg3z
      Preparing metadata (setup.py): started
      Preparing metadata (setup.py): finished with status 'error'
    
stderr:
      ERROR: Command errored out with exit status 1:
       command: /home/abc/.cache/pre-commit/repoyrc1wg3z/py_env-python3/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/abc/.cache/pre-commit/repoyrc1wg3z/setup.py'"'"'; __file__='"'"'/home/abc/.cache/pre-commit/repoyrc1wg3z/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-u6t1x1pm
           cwd: /home/abc/.cache/pre-commit/repoyrc1wg3z/
      Complete output (6 lines):
      Traceback (most recent call last):
        File "<string>", line 1, in <module>
        File "/home/abc/.cache/pre-commit/repoyrc1wg3z/setup.py", line 1
          from __future__ import annotations
          ^
      SyntaxError: future feature annotations is not defined
      ----------------------------------------
    WARNING: Discarding file:///home/abc/.cache/pre-commit/repoyrc1wg3z. Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
    ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

It now requires a minimum of Python 3.7 so we will need to use Conda to run the check. This adds a script to do that.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**To test:**

* Checking [testing job passed](https://builds.mantidproject.org/job/testing-pre-commit-script/15/)

*There is no associated issue.*

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

martyngigg committed:
Fix branch name master->main that had been missed!

martyngigg committed:
Add a conda-wrapper for the pre-commit script

pre-commit recently required Python 3.7+ so can no
longer run on stock CentoOS 7

martyngigg committed:
Fix check for presence of mambaforge installation

Removes additional argument to install script
that can be determined internally by the install
script

martyngigg committed:
Extract mamba setup to common function

Simplifies setting up a new environment & installing
packagines into it to 2 lines for most use cases.

sf1919 reviewed (dismissed):
Test instance passed no problems and pretty quickly.

Reviewing the code it's great to have the comments in to explain the changes.

Happy to approve.

thomashampson commented on buildconfig/Jenkins/Conda/publish-to-github:
```diff
@@ -29,20 +31,9 @@ github_release_cmd() {
-conda activate $CONDA_ENV_NAME
+# Mamba
+setup_mamba $WORKSPACE/mambaforge publish-gh
+mamba install gh
```
```suggestion
mamba install gh -y
```

thomashampson commented on buildconfig/Jenkins/Conda/download-and-install-mambaforge:
```diff
@@ -4,25 +4,28 @@
 # Expected args:
 #   1. EXPECTED_MAMBAFORGE_PATH: path to where mambaforge should be installed
-#   2. EXPECTED_CONDA_PATH: path to the conda executable
 #   3. CLEAN_BUILD: whether or not to force mambaforge to be removed before attempting to install it again
```
```suggestion
#   2. CLEAN_BUILD: whether or not to force mambaforge to be removed before attempting to install it again
```

martyngigg committed:
Ensure mamba does not prompt for input in scripts

--yes assumes yes for any questions that might be
asked about the installation.
